fdbclient/S3Client.actor.cpp

```diff
@@ -166,27 +351,207 @@ ACTOR Future<Void> copyUpBulkDumpFileSet(std::string s3url,
+		                                                                      objectName, // resource
+		                                                                      headers, // headers
+		                                                                      nullptr, // content
+		                                                                      0, // content length
+		                                                                      { 200 })); // expected response codes
+
+		auto it = response->data.headers.find("Content-MD5");
+		if (it != response->data.headers.end()) {
+			return Optional<std::string>(it->second);
+		}
+
+		return Optional<std::string>();
+	} catch (Error& e) {
+		// If we can't get the MD5, just return empty optional
+		TraceEvent(SevWarn, "S3ClientGetMD5Failed")
+		    .error(e)
+		    .detail("Bucket", bucket)
+		    .detail("Object", objectName)
+		    .detail("Offset", offset)
+		    .detail("Size", size);
+		return Optional<std::string>();
+	}
+}
+
 // Copy down file from s3 to filepath.
 ACTOR static Future<Void> copyDownFile(Reference<S3BlobStoreEndpoint> endpoint,
                                        std::string bucket,
-                                       std::string resource,
-                                       std::string filepath) {
+                                       std::string objectName,
+                                       std::string filepath,
+                                       PartConfig config = PartConfig()) {
 	TraceEvent("S3ClientCopyDownFileStart")
 	    .detail("filepath", filepath)
 	    .detail("bucket", bucket)
-	    .detail("resource", resource);
-	std::string content = wait(endpoint->readEntireFile(bucket, resource));
-	auto parent = std::filesystem::path(filepath).parent_path();
-	if (parent != "" && !std::filesystem::exists(parent)) {
-		std::filesystem::create_directories(parent);
+	    .detail("objectName", objectName);
+	state int64_t fileSize = wait(endpoint->objectSize(bucket, objectName));
+	state Reference<IAsyncFile> file;
+	state std::vector<PartState> parts;
+
+	try {
+		// Create or truncate the output file. Create subdirs if needed.
+		platform::createDirectory(parentDirectory(filepath));
+
+		// Open file asynchronously with write access
+		Reference<IAsyncFile> f = wait(IAsyncFileSystem::filesystem()->open(
+		    filepath,
+		    IAsyncFile::OPEN_ATOMIC_WRITE_AND_CREATE | IAsyncFile::OPEN_CREATE | IAsyncFile::OPEN_READWRITE,
+		    0644));
+		file = f;
+
+		// Pre-allocate file space if possible
+		wait(file->truncate(fileSize));
+
+		// Prepare part states
+		state int64_t offset = 0;
+		while (offset < fileSize) {
+			state PartState part;
+			part.offset = offset;
+			part.size = std::min(config.partSizeBytes, fileSize - offset);
+
+			// Try to get MD5 for this range
+			Optional<std::string> md5 = wait(getObjectRangeMD5(endpoint, bucket, objectName, part.offset, part.size));
+			if (md5.present()) {
```

Comment:
Similar to copyUp. Do we want to print some warning or throw some error? We may want users to enforce the MD5 check.

Reply:
This md5'ing in download is not in latest patch. See above for why.